### DISCUSSION TITLE
Best way to auto-load a custom plugin on a hosted version

### DISCUSSION THREAD

**Discussion by lcetinsoy** [jgraph/drawio#2516] [Q&A]
Hello everyone ! 

First thank your for your work. 
I have developped a custom plugin that I used to add on an hosted version of drawio (using the docker image). 

There are unfortunatly two drawbacks : 

- you have to manually add the plugin from the menu
- this methods no longer works due to content policy issue (my draw io is not served from the same subdomaine as the plugin file)

I would like to build a drawio docker image where my plugin is load by default.

I plan on doing the following : 
1. fork the drawio repository and add the plugin file in "src/main/webapp/plugins"
2. tell drawio to load my plugin 
3. fork the https://github.com/jgraph/docker-drawio and tell it to use the fork of drawio to build the docker image
4. push the docker image to a registry and use it

Would that be correct / optimal ?
I am not sure on how to do step 2 

Thanks for your help in advance ! 

In case I manage to do it, I'd be happy to do a documentation PR or edit the wiki so that other people can do it them selves

Best regards and happy new year ! 

Laurent

**Comment by davidjgraph** [comment]
You could try [using the](https://github.com/jgraph/docker-drawio) [preconfig.js](https://github.com/jgraph/drawio/blob/dev/src/main/webapp/js/PreConfig.js) to [define a configuration](https://www.diagrams.net/doc/faq/configure-diagram-editor) that sets your plugin to load by default. See "plugins".

**Reply by fabian-xentral** [reply]
Hi @davidjgraph , is there any update / further guidance you can provide on how to best make custom plugins usable with draw.io? We'd need to host this somehow for other users to also get access - it won't work with a pure local install.

We're currently creating pretty neat but complex system documentations with draw.io, using animations, interactive effects such as showing/hiding shapes based on tags, layers, etc - and I'd love to get a customised version of the props module (e.g. to allow for HTML formatted content) as well as to integrate live data (tapping into the overlay plugin suggested on https://www.linkedin.com/pulse/how-i-created-my-first-plugin-drawio-rado-balvan) and so much more.

Thank you so much!

BR
f

**Comment by mark-az** [comment]
Hi Laurent, I recently did very similar thing on the hosted app not the docker version.
Anyway, I added my custom plugins in the App.pluginRegistry and then I used the 'p' parameter to activate 'on-the-fly' my two plugins
I hope this can help you in some way
Marco

**Reply by fabian-xentral** [reply]
Hey @mark-az any chance you can share a bit more detail on how you added your custom plugin to App.pluginRegistry?

I mean I can access that on the fly via the console - but that doesn't allow me to load the plugin when viewing a file without messing with the console.

Thank you !!

**Reply by mark-az** [reply]
Hi Fabian, long time has passed and I am currently out from that project. Anyway, if I remember well,  I updated a source file, probably App.js, to add the name of my custom plugin at the end of the array that contains all plugins. The 'registry' is just an array of strings.  After this, you can use the url P parameter in the url that client use to open/access  your custo draw.io  installation to specify which plugin you'd like to activate. Hope this can help you.
Regards,
Marco
